Migrate to Java 21 FFM API

frankgrimes97 opened the pull request:
The project's minimium java release requirements moved from Java 17 to Java 21.

Package imports changed from `jdk.incubator.foreign` -> `java.lang.foreign` and a number of API changes were required.
Notably, the new `java.lang.foreign.Arena` class needs to be exposed when allocating native memory regions as it controls the scope and lifecycle of allocations.

Respective API docs are as follows:
 - https://docs.oracle.com/en/java/javase/17/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/package-summary.html
 - https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/foreign/package-summary.html

frankgrimes97 committed:
Migrate to Java 21 FFM API

The project's minimum Java release requirements moved from Java 17 to Java 21.

Package imports changed from `jdk.incubator.foreign` -> `java.lang.foreign`
and a number of API changes were required.
Notably, the new `java.lang.foreign.Arena` class needs to be exposed when allocating
native memory regions as it controls the scope and lifecycle of allocations.

Respective API docs are as follows:
 - https://docs.oracle.com/en/java/javase/17/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/package-summary.html
 - https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/foreign/package-summary.html

frankgrimes97 committed:
Update GHA workflows to Java 21

leerho commented on src/main/java/org/apache/datasketches/memory/Memory.java:
```diff
@@ -102,39 +106,12 @@ static Memory map(File file) throws IOException {
-   * @param scope the given ResourceScope.
-   * It must be non-null.
-   * Typically use <i>ResourceScope.newConfinedScope()</i>.
-   * Warning: specifying a <i>newSharedScope()</i> is not supported.
```
In cases where the user supplies the _Arena_ I suggest we add in the javadoc something to the effect:

> This Memory component is not thread-safe as it was designed to be used in environments where threading and parallel operation are handled at higher levels.  Specifying _global()_, _ofAuto()_ or _ofShared()_ Arenas could produce unpredictable results

frankgrimes97 replied:
If `Arena.ofConfined()` is the only one which makes sense and is applicable here, then perhaps the Arena shouldn't be exposed in the public API at all?

leerho commented on src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -83,54 +83,50 @@ static WritableMemory writableWrap(
   /**
-   * Maps the specified portion of the given file into Memory for write operations.
-   * @param file the given file to map. It must be non-null and writable.
+   * Maps the specified portion of the given file into Memory for write operations with a ResourceScope.
```
I think you meant _Arena_, but that isn't correct either.  The write operations of WritableMemory assume that the MemorySegment exists and has already been allocated using an Arena.  (I think I made that mistake too.)

leerho commented on src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -83,54 +83,50 @@ static WritableMemory writableWrap(
+   * @param byteOrder the byte order to be used.  It must be non-null.
+   * @return mapped WritableMemory.
+   * @throws IllegalArgumentException -- if file is not readable or writable.
+   * @throws IllegalArgumentException -- if file is not writable.
```
Good Catch!

leerho commented on src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -83,54 +83,50 @@ static WritableMemory writableWrap(
-   * @param scope the given ResourceScope.
-   * It must be non-null.
-   * Typically use <i>ResourceScope.newConfinedScope()</i>.
-   * Warning: specifying a <i>newSharedScope()</i> is not supported.
```
Similar warning as above.

leerho commented on src/main/java/org/apache/datasketches/memory/MemoryRequestServer.java:
```diff
@@ -42,24 +42,24 @@ public interface MemoryRequestServer {
-   * It must be non-null.
+   * @param arena the Arena to be used for the newly allocated memory. It must be non-null.
    * Typically use <i>ResourceScope.newConfinedScope()</i>.
    * Warning: specifying a <i>newSharedScope()</i> is not supported.
```
Add warning as above. Also replace _ResourceScope_

leerho commented on src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -194,12 +193,13 @@ static WritableMemory allocateDirect(
    * @return WritableMemory
```
Add warning as above, also ResourceScope  needs to be replaced at about L186

leerho commented on src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -194,12 +193,13 @@ static WritableMemory allocateDirect(
       long capacityBytes,
       long alignmentBytes,
-      ResourceScope scope,
+      Scope scope,
```
I don't think we need and Arena and a MemorySegment.Scope.

leerho reviewed (changes requested):
Wow! This represents a huge amount of work on your part!  Thank you! 
I also recognize some improvements you made along the way!  

I don't know if I caught all the issues, many of which were related to comments.   

This is a huge PR and when I merge I will be merging into a dedicated branch, which will make it easier for both of us to see and find some things and run tests as we go. I do suspect this will take several passes to review and get right.  --cheers!

frankgrimes97 committed:
Minor cleanup based on PR feedback

frankgrimes97 committed:
Improve temp file creation in test suite

This is to try to address file permission errors seen in some GitHub Action
Workflow runs on Windows:
  "[INFO]
   Error:  Failures:
   Error:    LeafImplTest.checkMapLeafs:118 » AccessDenied TestFile2.bin
   Error:    SpecificLeafTest.checkMapLeafs:108 » AccessDenied TestFile2.bin"

frankgrimes97 committed:
Improve temp file creation in test suite

This is to try to address following errors seen in some GitHub Action
Workflow runs on Windows:
  "Warning:  Files with unapproved licenses:
  force_original.txt"
